### ISSUE TITLE
Copona is the new platform

### ISSUE THREAD

**Issue by prhost** [copona/copona#38]
Always wanted to organize the structure of opencart, I think very bad, especially this business of catalog and admin, there is little reuse of code mainly in models and configs. I want to leave the Copona in a new organization, I know it's a bit radical, but if you really want to start a new platform totally uncoupled to opencart, this is going to be one of the big differences (:

Come on!

First my idea is to separate the core of Copona into a repository and treat it as a composer package.

Then separate `install/` into a repository and this `install` will be a new one, doing communication with Github and remotely downloading all the source of the project and installing, but that is subject for another issue.

The `system/storage/` folder is just for storing files, like a CDN, so let's keep it in the root, inside it put the folder `image/`

Would be like this:

```
storage/
    |-- cache/
    |-- image/
    |-- logs/
    |-- modification/
    |-- upload/
```

------------------------------------------------

Already the future extensions would be in own folder uses, thus:

```
extensions/
    |-- Copona/
            |-- NewSliderShow/
            |-- AdvencedFilters
            |-- Paypal/
                   |-- catalog/
                   |-- theme/
                   |-- assets/
                   |-- config/
                   |-- route.php
                   |-- Extension.php
                   |-- ...
```

------------------------------------------------

Settings would be by 3 ways:
1) By standard files inside the `config/` folder right in the root;

2) For an `.env` file defining environment variables, there is a library that makes this kind of configuration very legal https://github.com/vlucas/phpdotenv;

3) In `.env` you will have a for example `ENV=production` which besides setting the project environment it will define a folder inside `config/` with that name, for example:

```
config/
    |-- database.php
    |-- testing/
    |-- production/
            |-- database.php
    |-- develop/
```

This is to configure settings for multiple environments at the same time.

------------------------------------------------

The theme will be totally separate in a folder in the root called `themes/` there will be placed all the files belonging to the theme, including the `language/` may end up disagreeing but I think this belongs more to the theme than any other place, And also a class called `Theme.php` which can have theme definitions and make it even more modular in the future.

```
themes/
    |-- Copona/
             |-- assets/
             |-- language/
             |-- template/
             |-- Theme.php
```

------------------------------------------------

And last but not least, the `modules/` folder where the main opencart folders that are `admin/`, `catalog/` and `system/` will be. This folder `core/` will be manipulated only by the composer, being only reading folders for the developer, and can not be edited. The composer will control its versions and update the platform. So when you have some bugfix or minor improvements just give a `composer update copona/core` (:

```
modules/
    |-- catalog/
    |-- admin/
    |-- system/
    |-- migrations/
    |-- tests/
    |-- ...
```

Note The admin route will be configurable and no longer fixed by accessing a folder through the url. So if the developer wants to access by `panel/` or `backend/` he can easily configure this.

------------------------------------------------

In short:

```
config/
core/
extensions/
themes/
storage/
.env
index.pgp
```

I know this is a drastic modification and @arnisjuraga will be scared of it. But I've done a lot of this in my projects and I know it's possible, just want to let me help. Let's build a new ecommerce platform, let Opencart and Arastta outdated against our platform. And make tutorials to further facilitate the migration of extensions and themes. So, shall we go?

**Comment by prhost** [comment]
https://github.com/hassankhan/config

**Comment by prhost** [comment]
https://github.com/nikic/FastRoute

**Comment by prhost** [comment]
https://github.com/Respect/Validation

**Comment by arnisjuraga** [comment]
@prhost - you have a great ideas! But you are right, i "am" :) scared of this idea. Basically, because of to main scope and idea of Copona - as a Opencart "kinda" _successor_. We would like to welcome any Opencart developer, to use Copona as base for new projects. For that - we have to keep it "working". 

Benefits from that:
- we have _stable_ base, where do we start from
- it can be interesting/usable for every developer/user/ who has worked with Opencart.
- with all the problems Opencart has, it still has many benefits compared with other Ecommerces out there.
- while the code base is still similar to original source, we can use Opencart community source code for fixes, security etc.

At the moment, if we drop everything we have and start a "new" ECommerce - it completely looses the idea of Opencart as base. I am not saying that this is worng :). And to be honest - sooner or later the code base will be so different, that we will not be compatible with Opencart anyway. 
And at that moment it will become a "new platform" anyway.

Meanwhile - there are still many problems, issues and improvements, we are planning. And if we decide to drop everything and start from scratch - then Copona could become a "stable" release in ... I don't  know - 2 years? We are running copona on live dev projects. This is the way, how we can spend time for developent. Otherwise - it would be too slow, if we do it out-of-the work time.

And here we are - this could be my current position: we are gathering all the ideas for a future, and welcoming as many developers we can, to build a stable community. For that - we need up-and-running version of the system.

I have moved this to Wiki section. Still anyone can post comments here, I will keep the discussion organized in Wiki tree.
https://github.com/copona/copona/wiki/New-Opencart-(brainstorm)

**Comment by prhost** [comment]
I thought that was the main idea of ​​the copona, to be totally different from Opencart, like the arastta. I already did 80% of the proposal of the issue in my projects, I took little time, in the maximum 2 to 3 months. We can try to get a separate branch dev while keeping the current project in another, the new and bugfix adjustments that it has in the current one goes manually to the dev, keeping it synchronized with the master it while developing the new structure. Unfortunately anything we changed in the project loses much compatibility with the opencart extensions due to vqmod/ocmod. My idea is not to make it totally incompatible with opencart, but to add new improvements and new tools, improve the core behind the application and maintain maximum compatibility with opencart extensions, including making the extensions and themes migration tool. For example the idea of ​​routes, migration or ORM, the extension will support without them but if the developer has the need to use them it will be much easier.

We need to get out of the ordinary, see how many forks opencart has, why this? Because Daniel is rude and archaic, so many people like us, we have the brilliant idea of ​​doing an opencart with our improvements, but in the end, all the forks look similar, and at one time or another Daniel absorbs all the ideas and applies To opencart as if it were his. We need to evolve, this is over, I'm disgusted with `$this->load->model()`, or `ControllerFooBarLongNameCamelCase` or the final Class. For example mail with PHPmailer. It's cool, way better than the old opencart nail art. But because we do not make a Mail interface and anyone who wants to implement phpmailer, swiftmailer, mailchimp, or others, only implements the interface with some other class and builds their own mail easy and fast. Just changing which mail to use in the config. I really want to help improve Copona and make it different and better. But I'll be brave if I'm going to continue to look like Opencart by just doing a little different thing that Daniel does not accept or will take so long to implement.

**Comment by prhost** [comment]
PS. brainstorm in portuguese means also "Sudden mental disorder" hehehe

**Comment by arnisjuraga** [comment]
It was too late for me, and I forgot to write, but you mentioned it already - yes, we can start a new branch, and make it creatively. And let's make some huge changes in next few months! I am not against that. And believe me - I am not against any improvements.

**Comment by lucasjkr** [comment]
I would be wary of rushing too fast to drastically alter the structure of Copona away from Opencarts base structure. Reason being that we would then lose the use of many existing extensions, modules, VQMOD's, etc, and those are the things that add value to a lot of Opencart deployments.

Secondly, I don't think that language strings should be merged into templates. Then we'd be stuck with language-specific templates, rather than what OC has now, well-built templates can use any supported language.

As I mentioned in my other comment, I think it's VERY important that we move the project files out of the base directory.

`/`
`/migrations`
`/public`
`---/catalog`
`---/admin`
`config.example.php`
`composer.json`
`.env` (if this is chosen)
`LICENSE`
`README`
`.gitignore`

There is a lot of stuff that's currently exposed to the world, and that could cause problems down the road.

I think there is a lot of more intelligent stuff that can be done with the file structure down the road, but since this is forking off of Opencart, why not focus on incremental improvements for the time being (lots of good ideas in here), and save major changes that break compatibility with OC for a future 3.0 or 4.0 release?

**Comment by prhost** [comment]
@lucasjkr 
The idea of ​​the Copona in the future is to decouple from Opencart, and turn independent.

On the compatibility of extensions, there is going to be a very large break, the only difference is that extensions have been separated to begin modularization. Something that never existed in opencart, but la within `extensions /` will follow the same structure that it has today in opencart, avoiding the maximum of breaking of compatibility. Let's do some tutorials and make it easy to migrate the opencart extension to copona.

On moving to public, this is bad because there are some hosts that do not give you root access, only to public, so you would have to modify paths and remove public. We can leave everything in public and trust htaccess more, being able to block folders that should not be accessed, work that way and I was never hacked. But it is a legal feature to make available that the developer chooses to have the public

About the language in the template. It would not all be moved to the theme, but let the theme add and overwrite languages ​​if you need to, because it always has places in the theme that are different from the default theme, and need to have language of that, stay moving the language file into the opencart ends Ceasing to be modular.

**Comment by arnisjuraga** [comment]
we are starting to create a modification for Copona - "custom plugin", and now is the right moment to decide - where to "put it"

I mean, in a directory structure

there are always 2 types fo modules:
- ones, which are Theme Specific (usable in and only in current theme). E.g. - custom order process step1, step2 etc. - and can be implemented using some (already in an experiment) hooks for original controllers

- ones, which are Universal, but Should not be included in Copona core, e.g. Shipping and Payment modules.

How do you think - do they must and can be located in one location?

~~~
/controller/plugins/ ?
~~~

~~~
/plugins ?
/sytem/storage/plugins ?
~~~

**Comment by prhost** [comment]
Always put modules/plugins/extensions in the root, this facilitates that in the future we can change the core place, and other folders without touching the structure of the modules, also collaborates with the security, not being inside the core folders, avoids permission problem and others Types of access failure that the module can have, also to remove and add, does not run the risk of losing files or conflicting with something.

I think we should follow the structure I had suggested, use it to this day, and make it much easier for me.

extensions/
    |-- Copona/ -> [Vendor]
            |-- NewSliderShow/ -> [Extension Name]
            |-- AdvencedFilters
            |-- General/
            |-- Paypal/
                   |-- catalog/
                   |-- theme/
                   |-- assets/
                   |-- config/
                   |-- route.php
                   |-- Extension.php
                   |-- ...

This structure makes it easy to add composer, and manage the extensions by it.

It does not have much to do, to have types of plugins, a project design and other generics. Usually when I have a lot of change from just that project, I create a general extension, and I put it in, which is specific.

And we should call extensions, which is the term called in opencart currently, and not to confuse those still developing for opencart, and someday want to migrate to Copona.

**Comment by arnisjuraga** [comment]
I am on integrating this for now. Give me a message, if you have opinion on this:

all extensions are places in /extensions/ folder, and is loaded both in admin and catalog depending on settings). 

~~~
extensions/
    |-- Copona/ -> [Vendor]
         |-- NewSliderShow/ -> [Extension Name]
         |-- AdvencedFilters
         |-- General/
         |-- Paypal/
               |-- catalog/
               |-- theme/
~~~

**Comment by prhost** [comment]
I think it's great for a start, need to prevent Camel case in folders like this in the example, this is crucial for namespaces and compatibility between operating systems. The exact same would be to convert all folders to Camel case, but not for now.

**Comment by arnisjuraga** [comment]
At the moment, idea will be to crate an identical structure as it's for old modules, this helps a lot to use any Opencart modules with small modification. We are on this right now.

**Comment by prhost** [comment]
Yes, this is a good idea, for the time being loader.php to search for extension files before the core file, I did an ExtensionManager to manage all this, including work with cache because later on it ends up getting a bit Heavy this search, even more if using glob function.

**Comment by arnisjuraga** [comment]
The first look: a66baee05811601beac88e458e79af509d6da021
At the moment, only Shipping extensions can be moved to ```/extensions/Vendor/ExtensionName/...``` structure, because we need to modify all other extension controllers in admin. 

We are using ```glob``` at the moment. If Extensions dir becomes quite big, then there will be performance issues. That can be resolved, by adding exact Extension file location in settings uppon install, not traversing all ```/extensions/*/*/``` folders every time.

**Comment by prhost** [comment]
To do a mapping of the files the first time and to save it in cache. Leaving this work to the developer gets very tiring and makes maintenance difficult.

**Comment by arnisjuraga** [comment]
I am thinking of making a method, to return "extension file if exists",
this will be much more maintainable and usable. All requests at one place.
Catching then can be implemented into this method easily.

2017. gada 29. maijs 19:47, "Kallef Alexandre de Souza" <
notifications@github.com> rakstīja:

**Comment by prhost** [comment]
https://github.com/copona/copona/pull/147
https://github.com/copona/copona/pull/151
https://github.com/copona/copona/pull/152

**Comment by arnisjuraga** [comment]
It's not hard to have a project with "only few open issues" :), if someone is banning everyone, who has capability to detect and post them :D 
https://github.com/opencart/opencart/issues/6035#issuecomment-335174971

**Comment by prhost** [comment]
Wow, Daniel must be lonely, sad and friendless 😞
